plugins/api/src/switch_user.cpp

```diff
@@ -0,0 +1,253 @@
+            for (auto* host_ptr = zone_ptr->primaryServerHost; host_ptr; host_ptr = host_ptr->next) {
+                switch_user_or_disconnect(*host_ptr);
+            }
+
+            for (auto* host_ptr = zone_ptr->secondaryServerHost; host_ptr; host_ptr = host_ptr->next) {
+                switch_user_or_disconnect(*host_ptr);
+            }
+        }
+    } // update_server_to_server_connections
+
+    auto rs_switch_user(RsComm* _comm, SwitchUserInput* _input) -> int
+    {
+        try {
+            // Only administrators are allowed to invoke this API.
+            // We check the proxy user because it covers clients and server-to-server redirects.
+            // Remember, the client and proxy user information is identical until there's a redirect.
+            if (!irods::is_privileged_proxy(*_comm)) {
+                log_api::error(
+                    "Proxy user [{}] does not have permission to switch client user.", _comm->proxyUser.userName);
+                return SYS_PROXYUSER_NO_PRIV;
+            }
+
+            // Return immediately if the client did not provide non-empty strings for the
+            // username and zone.
+            if (const auto ec = check_input(_input); ec != 0) {
+                return ec;
+            }
+
+            const auto* local_zone = getLocalZoneName();
+
+            if (!local_zone) {
+                log_api::error("Could not get name of local zone.");
+                return SYS_INTERNAL_ERR;
+            }
+
+            namespace adm = irods::experimental::administration;
+
+            // This call covers the existence check too.
+            const auto user_type = adm::server::type(*_comm, adm::user{_input->username, _input->zone});
+
+            if (!user_type) {
+                log_api::error(
+                    "[{}#{}] is not a user in the local zone [{}].", _input->username, _input->zone, local_zone);
+                return CAT_INVALID_USER;
+            }
+
+            const auto* user_type_string = adm::to_c_str(*user_type);
+            auto& client = _comm->clientUser;
+
+            // Update the user identity associated with the RsComm.
+            std::strncpy(client.userName, _input->username, sizeof(UserInfo::userName));
+            std::strncpy(client.rodsZone, _input->zone, sizeof(UserInfo::rodsZone));
+            std::strncpy(client.userType, user_type_string, sizeof(UserInfo::userType));
+
+            // Set the appropriate privilege level based on whether the user is local or remote
+            // to this zone.
+            if (std::strncmp(_input->zone, local_zone, sizeof(SwitchUserInput::zone)) == 0) {
+                client.authInfo.authFlag =
+                    (*user_type == adm::user_type::rodsadmin) ? LOCAL_PRIV_USER_AUTH : LOCAL_USER_AUTH;
+            }
+            else {
+                client.authInfo.authFlag =
+                    (*user_type == adm::user_type::rodsadmin) ? REMOTE_PRIV_USER_AUTH : REMOTE_USER_AUTH;
+            }
```

Comment:
If the server-to-server property is not set, the plugin needs to update the proxy user info as well.

If the server-to-server property is set, then do not update the proxy user info because the proxy user is the iRODS user administering the server.

Reply:
done